core/libp2p/peer/protocol_repository.hpp

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_PROTOCOL_REPOSITORY_HPP
+#define KAGOME_PROTOCOL_REPOSITORY_HPP
+
+#include <vector>
+
+#include <gsl/span>
+#include <outcome/outcome.hpp>
+#include "libp2p/basic/garbage_collectable.hpp"
+#include "libp2p/peer/peer_id.hpp"
+#include "libp2p/peer/protocol.hpp"
+
+namespace libp2p::peer {
+
+  /**
+   * @brief Storage for mapping between peer and its known protocols.
+   */
+  class ProtocolRepository : public basic::GarbageCollectable {
+   public:
+    ~ProtocolRepository() override = default;
+
+    /**
+     * @brief Add protocols to a peer.
+     * param p peer
+     * @param ms list of protocols
+     * @return peer error, if no peer {@param p} found
+     */
+    virtual outcome::result<void> addProtocols(
+        const PeerId &p, gsl::span<const Protocol> ms) = 0;
+
+    /**
+     * @brief Removes protocols from a peer.
+     * @param p peer
+     * @param ms list of protocols
+     * @return peer error, if no peer {@param p} found
+     */
+    virtual outcome::result<void> removeProtocols(
+        const PeerId &p, gsl::span<const Protocol> ms) = 0;
+
+    /**
+     * @brief Get all supported protocols by given peer {@param p}
+     * @param p peer
+     * @return list of protocols (may be empty) or peer error, if no peer
+     * @param p} found
+     */
+    virtual outcome::result<std::vector<Protocol>> getProtocols(
```

Comment:
Can't a result to span be returned in order to reduce string copies?

Reply:
STD set is not contiguous storage, it can not be casted to span